crates/sui/tests/shared_objects_tests.rs

```diff
@@ -199,6 +202,60 @@ async fn call_shared_object_contract() {
     assert!(effects.dependencies.contains(&assert_value_mut_transaction));
 }
 
+#[sim_test]
+async fn access_clock_object_test() {
+    let mut gas_objects = generate_test_gas_objects();
+
+    // Get the authority configs and spawn them. Note that it is important to not drop
+    // the handles (or the authorities will stop).
+    let configs = test_authority_configs();
+    let _handles = spawn_test_authorities(gas_objects.clone(), &configs).await;
+
+    // Publish the move package to all authorities and get its package ID.
+    let package_id = publish_counter_package(gas_objects.pop().unwrap(), configs.validator_set())
+        .await
+        .0;
+
+    let clock_object_arg = ObjectArg::SharedObject {
+        id: SUI_CLOCK_OBJECT_ID,
+        initial_shared_version: SUI_CLOCK_OBJECT_SHARED_VERSION,
+        mutable: false,
+    };
+
+    let transaction = move_transaction(
+        gas_objects.pop().unwrap(),
+        "clock",
+        "get_time",
+        package_id,
+        vec![CallArg::Object(clock_object_arg)],
+    );
+    let start = SystemTime::now()
+        .duration_since(SystemTime::UNIX_EPOCH)
+        .unwrap();
+    let effects = submit_shared_object_transaction(transaction, configs.validator_set())
+        .await
+        .unwrap();
+    let finish = SystemTime::now()
+        .duration_since(SystemTime::UNIX_EPOCH)
+        .unwrap();
+    assert!(matches!(effects.status, ExecutionStatus::Success { .. }));
+
+    assert_eq!(2, effects.events.len());
+    let event = effects.events.get(1).unwrap();
+    let Event::MoveEvent { contents, .. } = event else { panic!("Expected move event, got {:?}", event) };
+
+    use serde::{Deserialize, Serialize};
+    #[derive(Serialize, Deserialize)]
+    struct TimeEvent {
+        timestamp_ms: u64,
+    }
+    let event = bcs::from_bytes::<TimeEvent>(contents).unwrap();
+
+    // Some sanity checks on the timestamp that we got
+    assert!(event.timestamp_ms >= start.as_millis() as u64);
+    assert!(event.timestamp_ms <= finish.as_millis() as u64);
```

Comment:
Will `sim_test` help make this test reliable?

Reply:
This should be reliable regardless of simtest (as long as time does not go backwards somehow)

Reply:
Unfortunately, that can happen, because `SystemTime` is not guaranteed to be monotonic.  From [the docs](https://doc.rust-lang.org/std/time/struct.SystemTime.html):

> Distinct from the [Instant](https://doc.rust-lang.org/std/time/struct.Instant.html) type, this time measurement is not monotonic. This means that you can save a file to the file system, then save another file to the file system, and the second file has a SystemTime measurement earlier than the first. In other words, an operation that happens after another operation in real time may have an earlier SystemTime!

So we should probably use `Instant` instead here.  The other thing that I was wondering about was whether `sim_test` plays with time at all.  I.e. if it stops time, then this test becomes meaningless, because all the times are equal to each other.  Counteracting that (or the non-deterministic version of this where the test is just so fast it happens in less than a millisecond) is a little more difficult, because it requires saying that `start < finish`, which would require a sleep, or some help from sim_test to shift time forward, to make it true.

Reply:
I think while yes, this can happen, I don't think this is important to fix in the unit test...
We are talking about test failure in case ntp causes significant time correction during this particular check. I would wait to see a test failure first

Reply:
Agreed, it's probably rare, but the fix (to use `Instant`) is also straightforward, right?  Even if the risk of time travel is low, I'm not seeing what value we get from using `SystemTime` instead of `Instant` for measuring time deltas.

Reply:
I don't see how I can use `Instant` instead of SystemTime here?

Reply:
I agree this is not a big deal, but you could take a base time using SystemTime at the beginning, and add a duration that you get from an Instant. Then it would be monotonic.

But I think we can leave this alone unless it turns out to be flakey.

Reply:
I am not sure this would fix the problem. If we assume time can go back during the test, then time on checkpoint might be even less then initial base time in your proposal

Reply:
Yeah... okay, that just clicked for me as well -- yes, agreed, not so straightforward.

What does this imply for how time behaves on the network though?  Does it mean that we can see this behaviour in the wild (in theory), or is there a mechanism to ensure that successive checkpoints occur at later timestamps?

Reply:
In the real network when such correction occurs we would see some delays in narwhal certificates (since narwhal does not vote for headers in the future and instead has a sleep waiting for header time to become current)